ggml-cuda.cu

```diff
@@ -8892,14 +8906,21 @@ static void ggml_cuda_mul_mat_id(const ggml_tensor * src0, const ggml_tensor * s
                 GGML_ASSERT(row_id >= 0 && row_id < n_as);
 
                 CUDA_CHECK(cudaMemcpyAsync(dst_original + i01*nb1, dst_contiguous + num_src1_rows*nb1,
-                                        nb1, cudaMemcpyDeviceToDevice, stream));
+                                        nb1, dst_kind, stream));
                 num_src1_rows++;
             }
         }
 
         ggml_cuda_pool_free(src1_contiguous, as_src1);
         ggml_cuda_pool_free(dst_contiguous,  as_dst);
     }
+
+    if (dst_original_backend == GGML_BACKEND_CPU) {
+        CUDA_CHECK(cudaStreamSynchronize(stream));
+    }
+
+    src1_row.backend = src1_original_backend;
+    dst_row.backend  =  dst_original_backend;
```

Comment:
This seems unnecessary.

Reply:
Unless I'm missing something `src1_row` and `dst_row` are the original ggml tensors. If their backends were permanently set to `GGML_BACKEND_GPU` then the code would for the second eval assume that the data is already in VRAM and not copy it from RAM.

Reply:
`src1_row` and `dst_row` are copies, not the original tensors.

Reply:
Okay, it seems I misinterpreted the code. When `src1` and `dst` are dereferenced they seem to be copied so the changes should not be permanent.